File: leetcode-practice-java/src/main/java/io/hank/leetcode/complexity/ComplexityProcessor.java
package io.hank.leetcode.complexity;

import java.lang.reflect.Method;
import java.util.Arrays;

public class ComplexityProcessor {
    private ComplexityProcessor() {
        throw new IllegalStateException("Utility class");
    }

    public static <T> void printComplexity(T obj) {
        try {
            Method[] methods = obj.getClass().getDeclaredMethods();
            Arrays.stream(methods).forEach((method -> {
                TimeComplexity timeComplexity = method.getAnnotation(TimeComplexity.class);
                SpaceComplexity spaceComplexity = method.getAnnotation(SpaceComplexity.class);

                if (timeComplexity != null || spaceComplexity != null) {
                    System.out.println("Method Name: " + method.getName());
                }
                if (timeComplexity != null) {
                    System.out.println("Time Complexity: " + timeComplexity.value());
                }
                if (spaceComplexity != null) {
                    System.out.println("Space Complexity: " + spaceComplexity.value());
                }
            }));
        } catch (Exception e) {
            e.printStackTrace();
        }
    }
}

File: leetcode-practice-kotlin/src/main/kotlin/io/hank/leetcode/kotlin/complexity/ComplexityProcessor.kt
package io.hank.leetcode.kotlin.complexity

import java.lang.reflect.Method

object ComplexityProcessor {
    inline fun <reified T : Any> printComplexity(obj: T) {
        try {
            obj::class.java.declaredMethods.forEach { method ->
                val timeComplexity = method.getAnnotation(TimeComplexity::class.java)
                val spaceComplexity = method.getAnnotation(SpaceComplexity::class.java)

                if (timeComplexity != null || spaceComplexity != null) {
                    println("Method Name: ${method.name}")
                    timeComplexity?.let { println("Time Complexity: ${it.value}") }
                    spaceComplexity?.let { println("Space Complexity: ${it.value}") }

                }
            }
        } catch (e: NoSuchMethodException) {
            e.printStackTrace()
        }
    }
}